crates/agent/src/publications.rs

```diff
@@ -338,9 +338,20 @@ impl PublishHandler {
                 .await
                 .context("rolling back to savepoint")?;
 
+            // Add built specs to the draft for dry runs after rolling back other changes that do
+            // not apply to dry runs.
+            specs::add_built_specs_to_draft_specs(&spec_rows, &build_output, txn)
+                .await
+                .context("adding built specs to draft")?;
+
             return Ok((row.pub_id, JobStatus::Success));
         }
 
+        // Add built specs to the live spec when publishing a build.
+        specs::add_built_specs_to_live_specs(&spec_rows, &build_output, txn)
```

Comment:
I'm going back and forth on whether this should be integrated with `update_expanded_live_specs()` above.

The broader thought process is that, if a collection is updated, a materialization which sources from the collection (and which inlines its CollectionSpec) should _also_ be updated even if that materialization wasn't a draft spec of the publication. For these kinds of direct adjacencies, we _do_ want built_specs of adjacent specs to be updated.

But, currently `expanded_rows` is too over-broad. What I'd like to do, as future work, is:
 * Ensure a `built_spec` is populated for every `live_spec` row.
 * Implement a DB RPC that can be used to dynamically resolve `CollectionSpec`'s using`built_spec`, implementing the [ControlPlane::resolve_collections](https://github.com/estuary/flow/blob/master/crates/validation/src/lib.rs#L45-L85) trait.
  * _Then_ relax the graph expansion logic to only consider direct adjacencies of captures and materializations, rather than also expanding through it to all _other_ bound collections. This will significantly narrow the set of "expanded" specs.
  
Assuming we've done this ☝️ then I believe we'd _definitely_ want to update the `built_spec` of every expanded spec, in the same way that we do update their build ID today.

I'm also not aware of a hard reason that we _couldn't_ update `built_spec` of all current `expanded_rows` today, but it could result in more churn given the degree of graph (over)-expansion that we do today, and thus I 🧇.